BuildMasterAutomation/Functions/ConvertFrom-BMOtterScriptExpression.ps1

```diff
@@ -0,0 +1,189 @@
+        else
+        {
+            $Value = $Value -replace '^@\('
+        }
+
+        $Value = $Value -replace '\)$'
+
+        $closesNeeded = 0
+        $failingSyntax = $false
+        # Splitting up array or map by comma and collecting into array.
+        $parsedItems = &{
+            $start = 0
+            $end = 0
+            foreach ($i in 0..($Value.Length - 1))
+            {
+                $char = $Value[$i]
+
+                if ($char -eq '(' -and ($Value[$i - 1] -eq '@' -or $Value[$i - 1] -eq '%'))
+                {
+                    $closesNeeded++
+                    continue
+                }
+
+                if ($char -eq ')' -and $closesNeeded)
+                {
+                    $closesNeeded--
+                    if ($i -ne ($Value.Length - 1))
+                    {
+                        continue
+                    }
+                }
+
+                if ($closesNeeded -or ($char -ne ',' -and $i -ne ($Value.Length - 1)))
+                {
+                    continue
+                }
+
+                $end = $i
+                $lengthOfSubstring = $end - $start
+
+                if ($i -eq ($Value.Length - 1))
+                {
+                    $lengthOfSubstring++
+                }
+
+                $entry = $Value.Substring($start, $lengthOfSubstring).Trim()
+                $start = $i + 1
+
+                if ($isVector)
+                {
+                    $converted = ConvertFrom-BMOtterScriptExpression -Value $entry
+
+                    if ($converted -is [array])
+                    {
+                        # PowerShell Array flattening is the bane of my existence...
+                        @{ 'Item' = $converted } | Write-Output
+                        continue
+                    }
+                    $converted | Write-Output
+                    continue
+                }
+
+                $firstColonIndex = $entry.IndexOf(':')
+                if ($firstColonIndex -lt 0)
+                {
+                    $failingSyntax = $true
+                    break
+                }
+
+                @{
+                    'Key' = ConvertTo-Int -Item $entry.Substring(0, $firstColonIndex).Trim()
+                    'Value' = ConvertFrom-BMOtterScriptExpression -Value $entry.Substring($firstColonIndex + 1).Trim()
+                } | Write-Output
+
+                $start = $i + 1
+            }
+        }
+
+        if ($closesNeeded -or $failingSyntax)
+        {
+            return $originalValue
+        }
+
+        if ($isVector)
+        {
+            # More copium for dealing with array flattening :((
+            $parsedItems =
+                $parsedItems |
+                ForEach-Object {
+                    if ($_ -isnot [hashtable] -or -not $_.ContainsKey('Item'))
+                    {
+                        return $_
+                    }
+                    return ,$_['Item']
+                }
+            return $parsedItems

```

Comment:
This can be done more simply by wrapping a recursive call in an array expression `@()` and prefixing a comma `,` before it to prevent PowerShell from unrolling the collection.

```powershell
if ($isVector)
{
    return ,@($parsedItems | ConvertFrom-BMOtterScriptExpression)
}
```